src/compact.jl

```diff
@@ -112,33 +121,37 @@ macro compact(fex, _kwexs...)
   # make strings
   layer = "@compact"
   setup = NamedTuple(map(ex -> Symbol(string(ex.args[1])) => string(ex.args[2]), kwexs))
-  input = join(fex.args[1].args, ", ")
+  input =
+      try
+          fex_args = fex.args[1]
+          isa(fex_args, Symbol) ? string(fex_args) : join(fex_args.args, ", ")
+      catch e 
+        @warn "Function stringifying does not yet handle all cases. Falling back to empty string for input arguments"
+        ""
+      end
   block = string(Base.remove_linenums!(fex).args[2])
 
   # edit expressions
   vars = map(ex -> ex.args[1], kwexs)
-  @gensym self
-  pushfirst!(fex.args[1].args, self)
-  addprefix!(fex, self, vars)
+  fex = supportself(fex, vars)
 
   # assemble
-  return esc(quote
-    let
-      $CompactLayer($fex, $name, ($layer, $input, $block), $setup; $(kwexs...))
-    end
-  end)
+  return esc(:($CompactLayer($fex, $name, ($layer, $input, $block), $setup; $(kwexs...))))
 end
 
-function addprefix!(ex::Expr, self, vars)
-  for i = 1:length(ex.args)
-    if ex.args[i] in vars
-      ex.args[i] = :($self.$(ex.args[i]))
-    else
-      addprefix!(ex.args[i], self, vars)
+function supportself(fex::Expr, vars)
+  @gensym self
+  @gensym curried_fex
+  # To avoid having to manipulate fex's arguments and body explicitly, we form a curried function first
+  # that wraps the full fex expression, and then uncurry it programatically rather than syntactically.
+  let_exprs = map(var -> :($var = $self.$var), vars)
+  return quote
+    $curried_fex = ($self) -> let $(let_exprs...) 
+        $fex
     end
+    ($self, args...; kwargs...) -> begin $curried_fex($self)(args...; kwargs...) end
```

Comment:
I was inspecting for macro hygiene... could someone with more experience tell me if I should actually gensym `args` and `kwargs` here?

Reply:
OK, I think it it's safe.